client/my-sites/sidebar/publish-menu.jsx

```diff
@@ -213,23 +223,30 @@ const PublishMenu = React.createClass( {
 
 		return (
 			<ul>
-				{ this.props.site && (
-					<QueryPostTypes siteId={ this.props.site.ID } />
+				{ this.props.siteId && (
+					<QueryPostTypes siteId={ this.props.siteId } />
 				) }
-				{ menuItems.map( this.renderMenuItem ) }
+				{ menuItems.map( this.renderMenuItem, this ) }
 			</ul>
 		);
 	}
-} );
+}
 
-export default connect( ( state ) => {
-	const siteId = get( getSelectedSite( state ), 'ID' );
+export default connect( ( state, { siteId } ) => {
 	const postTypes = getPostTypes( state, siteId );
 
 	return {
+		allSingleSites: areAllSitesSingleUser( state ),
+		canUser: ( cap ) => canCurrentUser( state, siteId, cap ),
```

Comment:
> I'd like us to avoid instantiating functions in `mapState`.

I was thinking the same thing. That was before I also tried to get rid of the `site` prop passed to `MediaLibraryUploadButton`, wrote 2 selectors to replace `media/util` helpers, and found that one of those helpers was used in another place, too.

> Would it be conceivable to gather all the items, pluck their `capability`, `uniq` that collection, and return a dictionary for the current site?

If anything, I think we shouldn't slice some data from `menuItems`, but rather do the entire filtering at `connect` stage.

But as indicated above, I'm a bit burnt out on this PR, so I wouldn't mind merging as-is. Ugliness aside, I don't think the function prop is going to cause any harm here. But I can file a TODO if you want.